crates/core/src/constraint_system/prove.rs

```diff
@@ -538,143 +547,186 @@ where
-
-				let selectors = flush
-					.selectors
-					.iter()
-					.map(|id| witness.get_multilin_poly(*id))
-					.collect::<Result<Vec<_>, _>>()?;
-
-				let n_vars = composite.n_vars();
-
-				let log_width = <PackedType<U, FExt<Tower>>>::LOG_WIDTH;
-
-				let len: usize = 1 << n_vars;
-				let packed_len: usize = 1 << n_vars.saturating_sub(log_width);
-
-				let inner_c = composite.c();
-
-				let zero_suffixes = count_zero_suffixes(&selectors);
-
-				for (zero_suffix, id, poly) in izip!(&zero_suffixes, inner_polys, selectors) {
-					let nonzero_scalars_prefixes = len.saturating_sub(*zero_suffix);
-
-					witness.update_multilin_poly_with_nonzero_scalars_prefixes([(
-						*id,
-						poly,
-						nonzero_scalars_prefixes,
-					)])?;
-				}
-
-				let polys = inner_polys
-					.iter()
-					.map(|id| witness.get_multilin_poly(*id))
-					.collect::<Result<Vec<_>, _>>()?;
-
-				let max_packed_zero_suffix =
-					zero_suffixes.into_iter().max().unwrap_or(0) >> log_width;
-
-				let mut composite_data = (0..packed_len.saturating_sub(max_packed_zero_suffix))
-					.into_par_iter()
-					.map(|i| {
-						let evals = polys
-							.iter()
-							.map(|poly| {
-								<PackedType<U, FExt<Tower>>>::from_fn(|j| {
-									let index = i << <PackedType<U, FExt<Tower>>>::LOG_WIDTH | j;
-									poly.evaluate_on_hypercube(index).unwrap_or_default()
-								})
-							})
-							.collect::<Vec<_>>();
-
-						inner_c
-							.evaluate(&evals)
-							.expect("query length is the same as poly length")
-					})
-					.collect::<Vec<_>>();
-
-				// `ArithExpr::Const(F::ONE) + selector * arith_expr_linear` — so if selector is
-				// zero, we fill with ones.
-				composite_data.resize(packed_len, ones);
-
-				let composite_poly = MultilinearExtension::new(n_vars, composite_data)
-					.expect("data is constructed with the correct length with respect to n_vars");
-
-				Ok((flush_oracle, MLEDirectAdapter::from(composite_poly).upcast_arc_dyn()))
-			}
-			MultilinearPolyVariant::LinearCombination(lincom) => {
-				let polys = lincom
-					.polys()
-					.map(|id| witness.get_multilin_poly(id))
-					.collect::<Result<Vec<_>, _>>()?;
-
-				let packed_len = 1
-					<< lincom
-						.n_vars()
-						.saturating_sub(<PackedType<U, FExt<Tower>>>::LOG_WIDTH);
-				let lin_comb_data = (0..packed_len)
-					.into_par_iter()
-					.map(|i| {
-						<PackedType<U, FExt<Tower>>>::from_fn(|j| {
-							let index = i << <PackedType<U, FExt<Tower>>>::LOG_WIDTH | j;
-							polys.iter().zip(lincom.coefficients()).fold(
-								lincom.offset(),
-								|sum, (poly, coeff)| {
-									sum + poly
-										.evaluate_on_hypercube_and_scale(index, coeff)
-										.unwrap_or(<FExt<Tower>>::ZERO)
-								},
-							)
-						})
-					})
-					.collect::<Vec<_>>();
+	let indices_to_update = flush_oracle_ids
+		.par_iter()
+		.zip(flushes)
+		.map(|(&flush_oracle, flush)| {
+			let oracle = oracles.oracle(flush_oracle);
+			let n_vars = oracle.n_vars();
```

Comment:
`oracle` does a clone of the oracle, it's expensive-ish and here we are only interested in `n_vars`. So choose the one below.

```suggestion
			let oracle = &oracles[flush_oracle];
			let n_vars = oracle.n_vars();
```

```suggestion
			let n_vars = oracles.n_vars(flush_oracle);
```